### PR TITLE
Separate S3 APIs from other multi-cloud APIs and change the authentification of S3 APIs to be compatible with AWS S3.

### LINKED ISSUE #893: S3 APIs are not AWS S3 compatible.
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug

**What happened**:
Two parts:
1. Currently, the URL of S3 APIs is like http(s)://ip:port/v1/s3/bucketname, which is not compatible with AWS S3, v1/s3/ should be removed.
At the same time, the authentification used a private algorithm, which is not compatible with AWS S3 too.
2. Each s3 API did not do AWS s3 compatilbe test, they may not be AWS S3 compatible.

Because of the above reasons, the third party s3 client, like the s3 browser, cannot be used to test multi-cloud.

**What you expected to happen**:
The third-party s3 client, like the s3 browser, cannot be used to test multi-cloud.

**How to reproduce it (as minimally and precisely as possible)**:
NA

**Anything else we need to know?**:

**Environment**:
- Gelato(release/branch) version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by sfzeng**
**What this PR does / why we need it**:
This PR is the first PR to make multi-cloud be compatible with AWS S3. Two things include in this PR:
1. Separate S3 APIs from other multi-cloud APIs.
2. Change the authentification of S3 APIs, make it support auth v2/v4, which is the same as AWS S3.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #893 

**Special notes for your reviewer**:
1. This is the first PR for AWS S3 compatible issue, there may have some duplicate code between s3api and api. In the next PR I will remove those duplicate code by creating some public folders to store those duplicate code.
2. This PR just fix the URL and authentification issues, but not make sure all S3 APIs be compatible with AWS S3. We will do AWS S3 compatible test next, and raise new PRs if needed.

**Release note**:

```
As S3 APIs are separated from other APIs, the dashboard needs to change accordingly. The dashboard will be changed later.
```

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -38,6 +38,24 @@ services:
...
+      - 8090:8090
+    environment:
+      - MICRO_SERVER_ADDRESS=:8090
+      - MICRO_REGISTRY=mdns
+      - OS_AUTH_AUTHSTRATEGY=keystone
+      - OS_AUTH_URL=http://10.10.3.151/identity
+      - OS_USER_DOMIN_ID=Default
+      - OS_USERNAME=opensds
+      - OS_PASSWORD=opensds@123
       - OS_PROJECT_NAME=service
```
Comment (sunfch) [inline docker-compose.yml]: missing  OS_PROJECT_NAME in s3api ?

File: s3api/cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	webService.Init()
+
+	obs.InitLogs()
+	wc := restful.NewContainer()
+
+	s3ws := new(restful.WebService)
+	s3ws.Path("/")
+	s3ws.Doc("OpenSDS Multi-Cloud API")
+	s3ws.Consumes(restful.MIME_XML)
+	s3ws.Produces(restful.MIME_XML)
```
Comment (sunfch) [inline s3api/cmd/main.go]: I dont know the  the reason for their existence of "s3ws.Consumes(restful.MIME_XML) and s3ws.Produces(restful.MIME_XML)",  and when I test s3 compatibility with this pr,  I found "s3ws.Consumes(restful.MIME_XML)" lead to error

File: s3api/pkg/filters/signature/signer/signature.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+			log.Errorf("[%s] reject: IsReqAuthenticated return false, err:%v\n", req.Request.URL, err)
+			s3.WriteErrorResponse(resp, req, err)
+			return
+		}
+		// TODO: check bucket policy
+	case AuthTypeAnonymous:
+		log.Errorf("[%s] reject: Anonymous not supported.\n", req.Request.URL)
+		s3.WriteErrorResponse(resp, req, s3error.ErrSignatureVersionNotSupported)
+		// TODO: check bucket policy
+		return
```
Comment (sunfch) [inline s3api/pkg/filters/signature/signer/signature.go]: i think it is wrong to return directly in the case

File: docker-compose.yml (hunk 1 above)
Comment (sfzeng) [reply docker-compose.yml]: Thanks, added.

File: s3api/cmd/main.go (hunk 2 above)
Comment (sfzeng) [reply s3api/cmd/main.go]: Could you please share what kind of error?

Comment (sfzeng) [reply s3api/cmd/main.go]: "Consumes specifies that this WebService can consume one or more MIME types. Http requests must have one of these values set for the Content-Type header."
Which kind of content type you used when error occured?

File: s3api/pkg/filters/signature/signer/signature.go (hunk 3 above)
Comment (sfzeng) [reply s3api/pkg/filters/signature/signer/signature.go]: Thanks, it's fixed.

File: s3api/cmd/main.go (hunk 2 above)
Comment (sfzeng) [reply s3api/cmd/main.go]: As this part is the same as before, and this PR is used to fix the issue that URL and authentication are not compatible aws s3. So can check and fix it in the following PR?

Comment (sunfch) [reply s3api/cmd/main.go]: http status code 415 was returned when I test s3 compatibility,  I commented "s3ws.Consumes(restful.MIME_XML)"  and the test gone on.  
I agree you that we check the issue in following PR.

**Comment by sfzeng** [comment]
Close it, will raise a new one.

**Pull request closed without merging**